autopilot/accounts.go

```diff
@@ -0,0 +1,201 @@
+	contracts, err := a.b.Contracts(ctx, cfg.Contracts.Set)
+	if err != nil {
+		a.logger.Errorw(fmt.Sprintf("failed to fetch contract set for refill: %v", err))
+		return
+	}
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	a.fundingContracts = append(a.fundingContracts[:0], contracts...)
+}
+
+func (a *accounts) refillWorkersAccountsLoop(stopChan <-chan struct{}) {
+	ticker := time.NewTicker(a.refillInterval)
+
+	for {
+		select {
+		case <-stopChan:
+			return // shutdown
+		case <-ticker.C:
+		}
+
+		a.refillWorkerAccounts()
+	}
+}
+
+// refillWorkerAccounts refills all accounts on a worker that require a refill.
+// To avoid slow hosts preventing refills for fast hosts, a separate goroutine
+// is used for every host. If a slow host's account is still being refilled by a
+// goroutine from a previous call, refillWorkerAccounts will skip that account
+// until the previously launched goroutine returns.
+func (a *accounts) refillWorkerAccounts() {
+	ctx, span := tracing.Tracer.Start(context.Background(), "refillWorkerAccounts")
+	defer span.End()
+
+	// Map hosts to contracts to use for funding.
+	a.mu.Lock()
+	contractForHost := make(map[types.PublicKey]api.ContractMetadata, len(a.fundingContracts))
+	for _, c := range a.fundingContracts {
+		contractForHost[c.HostKey] = c
+	}
+	a.mu.Unlock()
+
+	accounts, err := a.w.Accounts(ctx)
+	if err != nil {
+		a.logger.Errorw(fmt.Sprintf("failed to fetch accounts for refill: %s", err))
+		return
+	}
+	accountForHost := make(map[types.PublicKey]api.Account, len(accounts))
+	for _, acc := range accounts {
+		accountForHost[acc.Host] = acc
+	}
+
+	// Fund an account for every contract we have.
+	for _, contract := range contractForHost {
+		// Only launch a refill goroutine if no refill is in progress.
+		if !a.markRefillInProgress(contract.HostKey) {
+			continue // refill already in progress
+		}
+		go func(contract api.ContractMetadata) (err error) {
+			// Remove from in-progress refills once done.
+			defer a.markRefillDone(contract.HostKey)
+
+			// Fetch the account. This might be zero so use accordingly.
+			account := accountForHost[contract.HostKey]
+			if account.Balance == nil {
+				account.Balance = new(big.Int)
+			}
+			if account.Drift == nil {
+				account.Drift = new(big.Int)
+			}
+
+			// Add tracing.
+			ctx, span := tracing.Tracer.Start(ctx, "refillAccount")
+			defer span.End()
+			span.SetAttributes(attribute.Stringer("account", account.ID))
+			span.SetAttributes(attribute.Stringer("host", contract.HostKey))
+			span.SetAttributes(attribute.Stringer("balance", account.Balance))
+			defer func() {
+				if err != nil {
+					span.RecordError(err)
+					span.SetStatus(codes.Error, "failed to refill account")
+				}
+			}()
+
+			// Limit the time a refill can take.
+			ctx, cancel := context.WithTimeout(ctx, time.Minute)
+			defer cancel()
+
+			// Check if a host is potentially cheating before refilling.
+			if account.Drift.CmpAbs(maxDrift) > 0 {
```

Comment:
`maxDrift` or actually `maxNegDrift` since I renamed it is meant to be a barrier that requires manual intervention this time round.

e.g. if you download and fail due to the host reporting that there aren't enough funds and you check the account ans see that it got a `-10SC` drift associated with it, you can choose to reset it. Ideally we make sure that you never run into this as long as a host isn't cheating.